ksql-engine/src/main/java/io/confluent/ksql/materialization/Locator.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.materialization;
+
+import java.net.URL;
+import java.util.Optional;
+import org.apache.kafka.connect.data.Struct;
+
+/**
+ * Type used to locate on which KSQL node materialized data is stored.
+ *
+ * <p>Data stored in materialized stores can be spread across KSQL nodes. This type can be used to
+ * determine which KSQL server stores a specific key.
+ */
+public interface Locator {
+
+  /**
+   * Locate which KSQL node stores the supplied {@code key}.
+   *
+   * <p>Implementations are free to return {@link Optional#empty()} if the location is not known at
```

Comment:
> That's because we should always be able to determine where a key should be

That's simply not the case. During a partition rebalance there are times when a partition is not owned by any server.  During such times the location is _unknown_.

This is expected. 

A good programming matra is 'only through exceptions for exceptional cases'.  Not knowing the location is not exceptional, it's expected.  What's more, using `Optional` over an exception makes the fact that we can't always determine the owner a part of the API. Where as runtime exceptions are not part of the api.

Hence, IMHO, the signature of this method is correct.

```
Optional<KsqlNode> locate(Struct key);
```

It's saying 'call me to get the location of the key, and I'll return it if its known'.  It's then up to the caller to decide if it not being known is an error or not.

Reply:
That makes sense - I confused "knowing the partition" with "knowing the node"...